### DISCUSSION TITLE
Trying to place the popper horizontally at the click position

### DISCUSSION THREAD

**Discussion by sebastienbarre** [floating-ui/floating-ui#1382] [Q&A] [answered]
Our UI features a button that can (and will) grow significantly wide, since its label can be (and needs to be) very long.
Clicking on this button brings a pop-up menu below it, powered by Popper (thx).
However, this pop-up menu is currently centered with respect to the middle of the button.
```
              +------------------------------------------------------------------+
              |                            Long Button                           |
              +------------------------------------------------------------------+
                                                ^
                                     +--------------------+
                                     |       popper       |    
                                     +--------------------+
```

This makes sense for pretty much everything else in our UI that uses pop-up buttons, but in this case, if the user clicks near the edge of the button, they will have to travel a fair amount back to the center to reach the pop-up menu.

How can I write a modifier that changes the X position of the popper so that it matches the X position of the mouse where the user clicked on the button (that position will therefore be in the button, since they clicked on it). 

Could be near the left edge, could be near the right edge, could be 25% of the way, there is no fixed anchor, the anchor horizontally is the X position of the click event. For example:
```
              +------------------------------------------------------------------+
              |  click                     Long Button                           |
              +------------------------------------------------------------------+
                   ^
         +--------------------+
         |       popper       |    
         +--------------------+
```
or could also be:
```
              +------------------------------------------------------------------+
              |                            Long Button                   click   |
              +------------------------------------------------------------------+
                                                                           ^
                                                                  +--------------------+
                                                                  |       popper       |    
                                                                  +--------------------+
```
or anywhere horizontally, really:
```
              +------------------------------------------------------------------+
              |                            Long Button       click               |
              +------------------------------------------------------------------+
                                                               ^
                                                      +--------------------+
                                                      |       popper       |    
                                                      +--------------------+
```

Thank you

**Comment by FezVrasta** [comment, marked as answer]
It looks like you are looking for the top-start and top-end placements?

**Reply by sebastienbarre** [reply]
Thanks.
Did you mean `bottom-start` and `bottom-end` (the popper is below the button)? 
I'm afraid I don't see how that would work.
`top-start` would always place it near the left edge, and `top-end` near the right edge of the button, right? I just tried and that seems to be the case.

I'm not looking for a fixed anchor, I would like the popper to be horizontally centered with respect to the horizontal position of the *mouse click* (event X): 
* if they click near/at the center of the button, then the popper will be below the button, centered horizontally with respect to the button (as it usually is in UIs).
* if they click close to the left edge (because their mouse cursor was on the left of the screen for example, so the left edge of the button was the closest place to click), then the popper would be centered horizontally with respect to the click X position, meaning close to the left edge of the button. And vice-versa if they had clicked close to the right edge.

Either way edges don't matter here, essentially I would like the popper to show up below the button, but as close to the place where they clicked as possible, horizontally. Could be at the center, near the left edge, right edge, 25% of the way on the left edge, etc. This wouldn't matter usually because buttons are small, but this one happens to be very long :) 

Thank you

**Reply by FezVrasta** [reply]
My bad, I was from my phone and mixed up buttons and poppers.

Have you tried to call `setOptions()` in order to set the required offsets based on the click position to the `offset` modifier?

- https://popper.js.org/docs/v2/constructors/#setoptions
- https://popper.js.org/docs/v2/modifiers/offset/

**Reply by sebastienbarre** [reply]
@FezVrasta  No worries, I appreciate all the help. I can't find the documentation that shows how to access the event position though :(

**Reply by FezVrasta** [reply]
I suppose it would look like this:


```js
buttonElement.addEventListener('click', evt => {
   const rect = evt.target.getBoundingClientRect();
   const x = evt.clientX - rect.left; // x position within the element

   popperInstance.setOptions(options => ({
     ...options,
     modifiers: [
       ...options.modifiers,
       { name: 'offset', options: { offset: [x, 0] } },
     ],
   }));

});

```

**Reply by FezVrasta** [reply]
I prepared a Code Sandbox to show you how you could achieve that

https://codesandbox.io/s/prod-fast-vbxk8?file=/src/index.js

**Reply by sebastienbarre** [reply]
@FezVrasta I appreciate your patience. I'm using `usePopper()` though, and looking at the [source code](https://github.com/popperjs/react-popper/blob/master/src/usePopper.js#L145) it doesn't look like I have access to the instance -- should I use `update` or `forcedUpdate` instead somehow? Thx.

Thanks for the codesandbox. Now to try to adapt it to React-Popper...

**Reply by FezVrasta** [reply]
With React Popper you should store the offset in a React state, and then apply it to the options configuration passed to the `usePopper` hook.

Memoize (`useMemo`) the options config otherwise you'll end up in a loop.

```js

const [offset, setOffset] = useState(0);
const options = useMemo(() => ({
  placement: "bottom-start",
  modifiers: [
    { name: "flip", enabled: false },
    { name: "offset", enabled: true, options: { offset: [offset, 0] } },
  ]
}), [offset]);

const {} = usePopper(options);
```

**Reply by FezVrasta** [reply]
FYI I think Tippy.js provides this functionality out of the box.

**Reply by sebastienbarre** [reply]
Thank you very much
